Deprecate XX:IdleTuningCompactOnIdle

ShishirH opened the pull request:
Added dependency to XX:IdleTuningCompactOnIdle and
deprecated XX:IdleTuningCompactOnIdle. Also documented
added heuristics in version changelogs.

Fixes #277 

Signed-off-by: Shishir Halaharvi <shishir.neo95@gmail.com>

ShishirH commented:
Changes reflected in documentation:
1. `XX:IdleTuningCompactOnIdle` now has no effect when `XX:IdleTuningGcOnIdle` is not enabled.
2. Heuristics have been added to automatically compact the heap during idle GC when they have been triggered. As such enabling `XX:IdleTuningGcOnIdle` might also compact the heap. To force compact everytime idle GC is performed, `XX:IdleTuningCompactOnIdle` needs to be enabled.
3. Due to the heuristics being added, `XX:IdleTuningCompactOnIdle` has been deprecated.

ShishirH commented:
@SueChaplain @DanHeidinga Can you please review the changes and suggest corrections, if any?

ashu-mehra commented on docs/version0.15.md:
```diff
@@ -74,6 +74,9 @@ This option was redundant and has now been removed. If you try to use this optio
 `JVMJ9VM007E Command-line option unrecognised: -Xdiagnosticscollector`
 
+### Heuristics for compaction during idle GC
+OpenJ9 now automatically compacts the heap when certain triggers are met during idle GC. As a result of this change, [`-XX:IdleTuningCompactOnIdle`](xxidletuningcompactonidle.md) has been deprecated.
```
Should we be more specific and mention `-XX:[+|-]IdleTuningCompactOnIdle`.

ShishirH replied:
Fixed

ashu-mehra commented on docs/xxidletuningcompactonidle.md:
```diff
@@ -28,7 +28,11 @@
+1. This option is deprecated and will be removed in the future. 
+2. This option applies only to Linux&reg; architectures when the Generational Concurrent (`gencon`) garbage collection policy is in use.
+3. This option is not effective if the object heap is configured to use large pages.
+4. This option is not effective when XX:IdleTuningGcOnIdle is not set.
```
Same as above - using `-XX:[+|-]IdleTuningGcOnIdle` instead of `XX:IdleTuningGcOnIdle`.

ShishirH replied:
@ashu-mehra Should it be "This option is not effective if XX:+IdleTuningGcOnIdle is not set?"

ashu-mehra replied:
This option is not effective if `-XX:+IdleTuningGcOnIdle` is not specified`.

ShishirH replied:
Changed accordingly.

ashu-mehra commented on docs/xxidletuninggconidle.md:
```diff
@@ -39,7 +39,7 @@ This option controls whether a garbage collection cycle takes place when the sta
 | `-XX:-IdleTuningGcOnIdle` | Disable |  <i class="fa fa-check" aria-hidden="true"></i><span class="sr-only">yes</span>   |   |
 
-The default depends on whether or not the OpenJ9 VM is running in a docker container. As indicated in the table, when the VM is running in a container and the state is set to idle, this option causes the VM to release free memory pages in the object heap without resizing the Java&trade; heap. The pages are reclaimed by the operating system, which reduces the physical memory footprint of the VM. I ...
+The default depends on whether or not the OpenJ9 VM is running in a docker container. As indicated in the table, when the VM is running in a container and the state is set to idle, this option causes the VM to release free memory pages in the object heap and attempts to compact the heap after the garbage collection cycle if certain heuristics are triggered. The pages are reclaimed by the operatin ...
```
Previous content mentions `without resizing the Java&trade; heap.` which I think is important and should not be removed. Even if we do compaction, we won't be resizing the heap. So it still holds true.

ShishirH replied:
Made suggested change.

ashu-mehra commented:
If we are marking an option as deprecated, should the links to that option page be removed?

SueChaplain commented:
Jenkins doc stage

SueChaplain commented:
@ShishirH - thanks for addressing this one. I staged your changes here:
https://eclipse.github.io/openj9-docs-staging/314/version0.15/

A few comments:


- The What's new (version0.15.md) topic has a section but there is no linked entry in the list at the top.
- xxidletuninggconidle.md is updated with a deprecation statement but does not say which release it is deprecated in. Please can we add "deprecated in 0.15.0".  Also, the list of restrictions has wrapped into a line, probably a space missing somewhere in the markdown.
- is this still true of xxidletuningcompactonidle.md -> *If your application is not running in a container and you want to enable compaction as part of the idle-tuning process, set the -XX:+IdleTuningCompactOnIdle option on the command line when you start your application* With this change, would we expect *compact* to happen anyway? If the change is effective only for a container, I think the release topic should state this. Please clarify or explain to help me understand.

Thank you

SueChaplain commented:
Reading this again, is the *what's new* also missing one key fact - "IdleTuningCompactOnIdle is not effective if XX:+IdleTuningGcOnIdle is not specified"

ShishirH commented:
Thanks for the quick review @SueChaplain.

>  The What's new (version0.15.md) topic has a section but there is no linked entry in the list at the top.

Added the entry

>  xxidletuninggconidle.md is updated with a deprecation statement but does not say which release it is deprecated in. Please can we add "deprecated in 0.15.0".  

Changed statement to include deprecation release.

>Also, the list of restrictions has wrapped into a line, probably a space missing somewhere in the markdown.

Added whitespace. Can you please check if this is no longer wrapped now?

> is this still true of xxidletuningcompactonidle.md -> _If your application is not running in a container and you want to enable compaction as part of the idle-tuning process, set the -XX:+IdleTuningCompactOnIdle option on the command line when you start your application_ With this change, would we expect _compact_ to happen anyway? If the change is effective only for a container, I think the release topic should state this. Please clarify or explain to help me understand.

CompactOnIdle is enabled by default in containers. As such, compaction happens everytime idle GC happens in a container environment. It's true that outside containers, compaction might still happen even without specifying the CompactOnIdle option. 

Changed the statement to "If your application is not running in a container and you want compaction to be attempted every time idle GC happens as part of the idle-tuning process, set the `-XX:+IdleTuningCompactOnIdle` option on the command line when you start your application." to be more accurate.

> Reading this again, is the _what's new_ also missing one key fact - "IdleTuningCompactOnIdle is not effective if XX:+IdleTuningGcOnIdle is not specified"

Included this change in the release.

ashu-mehra commented:
Now that we are marking `-XX:+IdleTuningCompactOnIdle` as deprecated, do we need to remove it from AdoptOpenJDK-OpenJ9 Dockerfiles as well? @dinogun @DanHeidinga

DanHeidinga commented:
> Now that we are marking -XX:+IdleTuningCompactOnIdle as deprecated, do we need to remove it from AdoptOpenJDK-OpenJ9 Dockerfiles as well?

I think that would be wise.  We should be aiming for non-deprecated options in the Adopt dockerfiles so they serve as both a good base and a good example of how to configure openj9 in a docker

dinogun commented:
Yes, Will update our images, thanks

SueChaplain commented:
Jenkins doc stage

SueChaplain commented:
@ShishirH - thanks for making those changes and sorry it took me a while to get the changes staged. There are still a couple of formatting issues and the errors aren't obvious,  but I'm going to merge these changes and have a fiddle with the markdown to iron them out - save you needing to come back to this again.
